crate_universe/extension.bzl

```diff
@@ -225,7 +258,23 @@ def _crate_impl(module_ctx):
             annotations = {k: v for k, v in module_annotations.items()}
             for crate, values in repo_specific_annotations.get(cfg.name, {}).items():
                 _get_or_insert(annotations, crate, []).extend(values)
-            _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations)
+            cargo_lockfile = module_ctx.path(cfg.cargo_lockfile)
+            manifests = {str(module_ctx.path(m)): str(m) for m in cfg.manifests}
+            _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations, cargo_lockfile = cargo_lockfile, manifests = manifests)
+            all_repos.append(cfg.name)
+            local_repos.append(cfg.name)
+
+        for cfg in mod.tags.from_specs:
```

Comment:
Nit: Looks like a lot of duplicated code between this and the above section. Could we pull the common stuff into a function?

Reply:
good call, I rearranged this a bit to share the repo validation. So the duplication now is the 3 lines for `annotations`. I'm hesitating a bit because sharing this either requires using a lambda or a function that takes `module_annotations`, `repo_specific_annotations`, and `cfg.name`. Alternately, it can look like `this_repo_annotations = update_annotations(module_annotations, repo_specific_annotations.get(cfg.name, {})` or something. 

Do you still think it's worth pulling out or keep it as-is?

Reply:
I quite like the `repo_specific_annotations` idea. I think that'd probably be the cleanest.